Fix performance issues when moving the mouse with high polling rate on Windows

Calinou opened the pull request:
> [!NOTE]
> 
> **Testing needed on various hardware configurations:** V-Sync on/off, various refresh rates, pen tablet.
Make sure to build with optimizations enabled, or use the Windows release export template linked below.

This leads to a significant improvement on my end; aiming in a first-person game now looks pretty smooth in action when using a mouse with a high polling rate (>= 2000 Hz).

On my own setup with an optimized editor binary, I get these FPS figures when running InputRepro2 with `--disable-vsync` and moving a 8,000 Hz mouse as quickly as possible:

Maximized window:

- Not moving the mouse: 3735 FPS (0.27 mspf)
- Mouse mode other than captured (input accumulation enabled/disabled, same performance): 902 FPS (1.11 mspf)
- Captured mouse mode (input accumulation enabled): 1460 FPS (0.68 mspf)
- Captured mouse mode (input accumulation disabled) 1280 FPS (0.78 mspf)

Fullscreen:

- Not moving the mouse: 3815 FPS (0.26 mspf)
- Mouse mode other than captured (input accumulation enabled/disabled, same performance): 1058 FPS (0.95 mspf)
- Captured mouse mode (input accumulation enabled): 1604 FPS (0.62 mspf)
- Captured mouse mode (input accumulation disabled) 1432 FPS (0.70 mspf)

The demo remains CPU-limited even when not moving the mouse. Interestingly, a maximized window runs slightly slower than fullscreen when moving the mouse. Using a release export template binary scores ~10% higher framerates across the board (except when not moving the mouse, where it's mostly identical).

**Testing projects:**

- [InputRepro2.zip](https://github.com/user-attachments/files/22754573/InputRepro2.zip)
 (general framerate stability testing)
- [graphics_tablet_input.zip](https://github.com/user-attachments/files/21798498/graphics_tablet_input.zip) (to test pen tablets, but also drawing precision with a mouse)
- [physics_interpolation.zip](https://github.com/user-attachments/files/21798491/physics_interpolation.zip) (to test captured mouse mode FPS look)

Windows x86_64 binaries for testing (compiled with MSVC 2022 and `production=yes`):

- [Editor](https://github.com/Calinou/media/releases/download/download/godot.windows.editor.x86_64.gh-109639.zip)
- [Release export template](https://github.com/Calinou/media/releases/download/download/godot.windows.template_release.x86_64.gh-109639.zip)

___

- This closes https://github.com/godotengine/godot/issues/60646.

Thanks to @bruvzg and @vorvek for helping out on this PR :slightly_smiling_face:

## TODO

- [x] I noticed an issue with captured mouse mode. In the InputRepro2 testing project, switch to captured mouse mode and click (or scroll the mouse wheel) while moving the mouse. You will see the mouse cursor appear to move for one frame, even though it shouldn't as we're in captured mouse mode. ~~This issue doesn't occur on `master`.~~
  - This still occurs on the latest iteration of this PR. However, testing as of June 2026 has found that this also occurs on `master` (it's harder to trigger and is less visible when it occurs though). That said, since mouse `position` is generally not relied upon when in captured mouse mode (`relative`/`screen_relative` is used instead), it's probably benign in practice.
- [x] Fix the issue mentioned in https://github.com/godotengine/godot/pull/109639#issuecomment-3378842876. It's possibly related to the item above.

## Preview

<details open>

<summary>PC specifications</summary>

- **CPU:** AMD Ryzen 9 9950X3D
- **GPU:** NVIDIA GeForce RTX 5090
- **RAM:** 64 GB (2×32 GB DDR5-6000 CL30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Windows 11 24H2
- **Monitor:** ASUS ROG PG32UCDP (480 Hz mode)
- **Mouse:** ASUS ROG Harpe Ace Extreme (2.4 GHz wireless, 8,000 Hz polling rate)
</details>

*Look not only at the FPS counter, but also at the mspf (frametime counter) next to it. We want it to be as stable as possible (as close as 2.1 mspf for a 480 Hz display).*

*The issue in `master` tends to be worse at lower FPS caps/refresh rates, but I wanted to test this PR in extreme scenarios to make sure as few frame drops as possible occur at very high refresh rates.*

### Before

#### Hidden mouse mode, input accumulation ON

https://github.com/user-attachments/assets/32baae7d-fb6e-4117-87ea-8dcfe1a2f6f5

#### Hidden mouse mode, input accumulation OFF

https://github.com/user-attachments/assets/340fcf20-13a0-4776-8f29-28f1c50662a2

#### Captured mouse mode, input accumulation ON

https://github.com/user-attachments/assets/68f00656-3226-43e5-83fc-71a74c1d795b

#### Captured mouse mode, input accumulation OFF

https://github.com/user-attachments/assets/77e232d4-0445-497b-bfd0-618d03b42ce1

### After

#### Hidden mouse mode, input accumulation ON

https://github.com/user-attachments/assets/8353715b-b25c-4e05-a49c-75b9178c3478

#### Hidden mouse mode, input accumulation OFF

https://github.com/user-attachments/assets/a81827ec-fa77-4e73-8d8a-b75970f279ab

#### Captured mouse mode, input accumulation ON

https://github.com/user-attachments/assets/d8be089b-fd60-4e28-a67f-1824300f86ea

#### Captured mouse mode, input accumulation OFF

https://github.com/user-attachments/assets/86045ad4-cbf1-44a4-bb22-3a88cd0591a4

Calinou committed:
Fix performance issues when moving the mouse with high polling rate on Windows

Co-authored-by: Pāvels Nadtočajevs <7645683+bruvzg@users.noreply.github.com>
Co-authored-by: Jon Tamayo <volsungn@gmail.com>

Calinou commented:
> - Non-mouse events (like keypresses)

This is something I noticed if we limit the number of processed events per frame to 1. We should make it so only mouse motion events are considered in the limit, but I'm not sure how.

Calinou commented:
> @Calinou , I have downloaded and extracted your export template, but after I rename it to my project and try to run it, it crashes. May I ask, what do I have to do to actually implement the fix?

You could try to compile the editor from source from this pull request's branch: https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_windows.html

You can get the source code of this branch here: https://github.com/Calinou/godot/archive/refs/heads/windows-fix-high-mouse-polling-rate-performance.zip

IneedHelp commented:
@Calinou thank you for the reply and information. Could you please tell me if the fix will eventually get to be implemented in an official release?

firebelley commented:
Hello, this issue has become a big problem in my game to the point where the game is almost unplayable. I just ran a test with the provided project and export template and was able to reproduce the issue and fix successfully.

I would love to test with my game, but unfortunately it's C# so I'd need to go through the process of building this branch from source with C# glue/bindings and that's going to be a multi-hour project.

Here is my hardware configuration:

- CPU: Intel(R) Core Ultra 9 285K (24 CPUs), ~3.7GHZ
- GPU: NVIDIA GeForce RTX 5080
- RAM: 64GB (2 x 32GB) DDR5-6400 PC5-51200 CL32
- SSD: Samsung 4TB 990 Pro NVME
- OS: Windows 11 Pro 64-bit (10.0, Build 26100)
- Monitor: Dell S2417DG at 144 Hz, GSync OFF
- Mouse: Keychron M6 at 8000hz polling rate

Below are my test results showing a big improvement on 8,000hz polling rate. Interestingly, I normally use the mouse at 1000hz and the problem seems to unreliably occur. Meaning that at 1,000hz polling rate, sometimes the game runs smooth and other times there is noticeable lag. Setting the mouse to 8,000hz is a surefire way to reproduce though.

<details>
  <summary>InputRepro2 on 4.5 stable, VSYNC ON</summary>
   

https://github.com/user-attachments/assets/feb1760a-1f1b-4114-91b6-54a82da9c734




</details>

<details>
  <summary>InputRepro2 with provided build, VSYNC ON</summary>


https://github.com/user-attachments/assets/b3a9d2f1-b4a3-4bd1-af4d-7c9be7c52b44


</details>

<details>
  <summary>InputRepro2 on 4.5 stable, VSYNC OFF</summary>


https://github.com/user-attachments/assets/a3502a88-0de0-4cfc-9ff5-8c9beaffe11d



</details>

<details>
  <summary>InputRepro2 with provided build, VSYNC OFF</summary>


https://github.com/user-attachments/assets/d665c8b6-edd9-4e6c-81b7-ed8ad9462ac3



</details>

Calinou commented:
Rebased and tested again, it works as expected. I've updated the first MRP to implement @firebelley's suggested peak mspf counter.

However, there's an issue I'm not sure how to fix. When you move the mouse quickly, some unrelated events like keyboard presses will act strangely, repeating themselves on certain occasions. This is noticeable in the `InputRepro2` testing project if you press <kbd>X</kbd> to toggle input accumulation while moving the mouse. The issue will occur more often if you lower `MAX_EVENTS_PER_FRAME` in `display_server_windows.cpp`. Setting it to `1` and running the project with **Max FPS** set to `10` in the Project Settings is a surefire way to making it happen.

In this case, holding down <kbd>X</kbd> will toggle input accumulation every frame, and if you moved the mouse after this, it will keep moving on its own as you press and release a key. I'm not sure how we can distinguish keyboard from mouse input in the function that calls `TranslateMessage()` and `DispatchMessageW()`.

@bruvzg Any ideas?

vorvek commented:
I tried to merge this into my Godot fork and ran into the captured-cursor behavior you mention at the end of the description. In my case it was not benign: it broke editor value drags and 3D viewport freelook.

The cause is the synthetic motion event added in `_set_mouse_mode_impl` when the mouse is captured:

```
mm->set_relative(Vector2(mm->get_position() - Vector2(old_x, old_y)));
mm->set_relative_screen_position(mm->get_relative());
old_x = mm->get_position().x;
old_y = mm->get_position().y;

mm->set_position(center);
```
`set_position(center)` runs after `set_relative`, so `mm->get_position()` is still `(0, 0)` on that line. The relative ends up as `(0, 0) - (old_x, old_y)`, the negated previous cursor position, which is a large delta on the first captured frame.

The cursor position blip is invisible while captured, like you noted. The part that bites is that captured-mode consumers read the relative motion, not the position. So an `EditorSpinSlider/SpinBox` drag jumps straight to its min or max, and the editor freelook camera snaps fully up or down the instant the drag starts.

The warp to the window center is a programmatic move, not real input, so the relative should be zero. This is what fixed it for me:

```
mm->set_position(center);
mm->set_global_position(center);
mm->set_relative(Vector2(0, 0));
mm->set_relative_screen_position(Vector2(0, 0));
mm->set_velocity(Vector2(0, 0));
mm->set_screen_velocity(Vector2(0, 0));

old_x = center.x;
old_y = center.y;
```
Resetting `old_x/old_y` to the center also gives the `MOUSE_MOVE_ABSOLUTE` branch in `process_raw_input()` a correct baseline, so absolute-reporting devices start from the center instead of a stale position. After this, drags and freelook start from rest.

Calinou commented:
@vorvek Thanks for testing and investigating the captured cursor issue! I've integrated the fix in the PR, and it seems to work better overall (tested on 1,000 Hz and 8,000 Hz polling rates). That said, the cursor will still sporadically move when clicking or using the mouse wheel in captured mouse mode (but at least the hardware cursor isn't visible anymore).

Freelook now works in the 3D editor, but only when holding the right mouse button. Pressing <kbd>Shift + F</kbd> to toggle freelook will instantly exit freelook due to the key combination being triggered twice in a row.

vorvek commented:
@Calinou Ah! Yes, I did have that issue too! The problematic part is the throttled Windows message pump in `process_events()`.

`TranslateMessage()` can post a `WM_CHAR` after a `WM_KEYDOWN`.  `MAX_EVENTS_PER_FRAME_KMB = 1` can dispatch the `WM_KEYDOWN` for `F` in one frame, then leave the generated `WM_CHAR` for the next frame.

Godot’s `_process_key_events()` only ignores that `WM_CHAR` when it is in the same `key_event_buffer` batch as the preceding keydown. If the `WM_CHAR` arrives alone on the next frame, this condition treats it as a fresh pressed key event:

```cpp
if (!KeyMappingWindows::is_extended_key(ke.wParam) && ((i == 0 && ke.uMsg == WM_CHAR) || (i > 0 && key_event_buffer[i - 1].uMsg == WM_CHAR))) {
```

I fixed it by keeping the raw input batching, but restoring normal Windows message queue draining after `process_raw_input()`:

https://github.com/godotengine/godot/blob/9d03a4937bf37c7c5719cbd79d0d94fbe59f0dcf/platform/windows/display_server_windows.cpp#L4235
```cpp
	_THREAD_SAFE_LOCK_

	process_raw_input();

	MSG msg = {};
	while (PeekMessageW(&msg, nullptr, 0, 0, PM_REMOVE)) {
		TranslateMessage(&msg);
		DispatchMessageW(&msg);
	}
	_THREAD_SAFE_UNLOCK_
```

That keeps the generated `WM_CHAR` in the same `_process_key_events()` batch as the `WM_KEYDOWN`, so it gets ignored as intended and `Shift + F` only toggles once.

If this PR still needs to avoid processing `WM_INPUT` through `PeekMessage()`, I think the important rule is: keyboard messages and any `WM_CHAR` generated by `TranslateMessage()` must not be split across frames before `_process_key_events()` runs. Draining all non-`WM_INPUT` messages after `process_raw_input()` should preserve the high-polling raw input benefit without breaking shortcuts.

vorvek commented:
I re-checked and noticed some other issues that weren't covered by the PR:

1) **The buffered read doesn't drain the queue.** With a null buffer, `GetRawInputBuffer()` returns the byte size of the *first* pending message in `pcbSize`, not a message count ([docs](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getrawinputbuffer)). So `dw_size = n_buffer * sizeof(RAWINPUT)` holds roughly `sizeof(RAWINPUT)` messages (about 48 mouse events on x64), and `process_raw_input()` only reads once per frame. An 8,000 Hz mouse at 60 FPS queues ~133 events per frame, so there's a leftover every frame during fast movement. In this PR the leftovers survive to the next frame because the `PeekMessage()` filters skip `WM_INPUT`, but they pile up as a backlog that adds latency exactly when the mouse moves fast. This is more or less the backlog @KeyboardDanni was worried about, just caused by the buffer sizing rather than the pump cap. The docs say to call it "several times ... until all raw input messages have been read", so I wrapped the read in a loop that runs until the queue reports empty.

On my fork it was worse than latency: I pump the queue unfiltered, so the leftover `WM_INPUT` messages got dispatched into the now-empty `WM_INPUT` case and freed unread. That dropped most of the captured-mode motion in the 8k @ 60 FPS scenario. Easy to miss at 1,000 Hz because everything fits in a single read.

2) **The no-op `WM_INPUT` case can still eat events.** Anything that arrives between the `GetRawInputBuffer()` call and the end of the message pump gets dispatched to the empty `WM_INPUT` handler, and `DefWindowProc` frees the data unread. In this PR that only happens through the unfiltered `PeekMessage()` used when `app_focused` is false, so it's rarer, but it's the same kind of loss. I restored the old `GetRawInputData()` handling there as a fallback. It only sees the handful of stragglers the batch read missed, so the per-message flood this PR fixes can't come back through it.

3) **WOW64.** The same docs page has a note that a 32-bit process on 64-bit Windows has to use `*pcbSize * 8` and align the `RAWINPUT` structures by 8 bytes, otherwise the buffer walk reads garbage. Doesn't affect me since my fork only ships x86_64, but the official x86_32 Windows templates would hit it, so it probably needs handling before this can merge.

4) **`process_raw_input()` hardcodes `MAIN_WINDOW_ID`**, while `_set_mouse_mode_impl()` captures the mouse for `_get_focused_window_or_popup()`. If a secondary window captures the mouse, the recentering and the relative math run against the wrong window. The old `WM_INPUT` path used the actual receiving window, so this is a small regression. I switched the buffered path to `_get_focused_window_or_popup()` with a main-window fallback.

I've opened Calinou/godot#20 against this PR's branch with the fixes for 1, 2 and 4.

Calinou commented:
This PR is ready for a further round of testing. Thanks to vorvek's help, it's now fully stable on all test projects, but I would like as many people as possible to test this on their own game projects. See OP for precompiled Windows binaries.

bruvzg reviewed (approved):
Code looks good, and I have not encountered any issue while testing. I was not able to reproduce huge performance drops.

Repiteo reviewed (approved):
Tested locally; everything checks out!

Repiteo commented:
Thanks!

Repiteo commented:
Cherry-picked for 4.7.2.

OwensC commented:
Thank you! This has been driving me nuts.

Repiteo commented:
Cherry-picked for 4.6.4.

thygrrr commented:
Just wanted to say that this issue was haunting me for over 2 years now and I thought it was a Godot problem, now it turns out it's a Godot solution.
